hooks: rework matplotlib.backends hook (#6024)
pyinstaller/pyinstaller: doc/hooks-config.rst

                 },
                 ...)

.. _matplotlib hook options:

Matplotlib hooks
----------------

The hooks for ``matplotlib`` package allow user to control the backend
collection behavior via ``backends`` option under the ``matplotlib``
identifier, as described below.

**Hook identifier:** ``matplotlib``

**Options**

 * ``backends`` [*string* or *list of strings*]: backend selection method
   or name(s) of backend(s) to collect. Valid string values: ``auto``,
   ``all``, or a human-readable backend name (e.g., ``TkAgg``). To
   specify multiple backends to be collected, use a list of strings
   (e.g., `['TkAgg', 'Qt5Agg']`).

**Backend selection process**

If ``backends`` option is set to `auto˙ (or not specified), the hook
performs auto-detection of used backends, by scanning the code for
:func:`matplotlib.use` function calls with literal arguments

> (e.g., ``matplotlib.use('TkAgg')`` beign used in the code results in

Reviewer: ```suggestion
(e.g., ``matplotlib.use('TkAgg')`` being used in the code results in
```

> ``TkAgg`` backend to be considered for collected).  If no such calls are

Reviewer: ```suggestion
the ``TkAgg`` backend to be considered for collection).  If no such calls are
```

present, the default backend is obtained by calling the

> :func:``matplotlib.get_backend`` function.

Reviewer: ```suggestion
:func:`matplotlib.get_backend` function.
```

If ``backends`` option is set to `all`, all (importable) backends are

> selection, which corresponds to the old behavior of |PyInstaller| 4.4

Reviewer: ```suggestion
selected, which corresponds to the old behavior of |PyInstaller| 4.4
```

and earlier. The list of importable backends depends on the packages
installed in the environment (e.g., ``Qt5Agg`` backend becomes importable
if environment has ``PyQt5`` or ``PySide2`` package installed).

> Otherwise, the value of ``backends`` option is treated as a backend

Reviewer: ```suggestion
Otherwise, the value of the ``backends`` option is treated as a backend
```

name (if it is a string) or a list of backend names (if it is a list).

In all cases, the outlined methods serve to obtain a list of *potential*
backends to collect, and only backends that can be imported (i.e., have
their dependencies installed) are actually collected.

**Example**

.. code-block:: python

    a = Analysis(["my-matplotlib-app.py"],
                 ...,
                 hooksconfig={
                     "matplotlib": {
                         "backends": "auto",  # auto-detect; the default behavior
                         # "backends": "all",  # collect all backends
                         # "backends": "TkAgg",  # collect a specific backend
                         # "backends": ["TkAgg", "Qt5Agg"],  # collect multiple backends
                     }
                 },

>                  ...)

Reviewer: I guess we should start wrapping example specs in the same format as #6137?

Author: That's a good idea. But to keep things consistent, the existing example specs in that file need to be rewrapped first..

Reviewer: I guess rewrap the rest in #6137 then do this example in the new style here.
